modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,197 @@
+                        .left
+                        .map(reason => RewardsException(reason))
+                  }
+                }.leftMap(_.asInstanceOf[RewardsError])
+              } yield (weightedDistribution.toSortedMap, SortedSet.from(transactions))
+          )
+
+      private[rewards] def getRewardsPerSnapshot(
+        snapshotOrdinal: SnapshotOrdinal,
+        timeTriggerInterval: FiniteDuration
+      ): Either[RewardsError, Amount] =
+        for {
+          snapshots <- getSnapshotsCountPerEpoch(timeTriggerInterval)
+          epoch <- getEpochNumber(snapshotOrdinal, snapshots)
+          reward <- getTotalRewardsPerEpoch(epoch)
+          // Note: We need rounding there to match validator rewards sheet rounding
+          rewardPerSnapshot <- NonNegLong
+            .from(Math.round(reward.value / snapshots.toDouble))
+            .left
+            .map(RewardsException)
+        } yield Amount(rewardPerSnapshot)
+
+      private[rewards] def getSnapshotsCountPerEpoch(
+        timeTriggerInterval: FiniteDuration
+      ): Either[RewardsError, PosLong] =
+        PosLong
+          .from((config.epochDurationInYears * 12 * 30 * 24 * 60).toLong / timeTriggerInterval.toMinutes)
+          .left
+          .map(RewardsException)
+
+      private[rewards] def getTotalRewardsPerEpoch(epochNumber: PosInt): Either[RewardsError, Amount] =
+        NonNegLong
+          .from(
+            (config.baseEpochReward.value / Math.pow(2, epochNumber.toDouble - 1.0)).toLong
+          )
+          .map(Amount(_))
+          .left
+          .map(RewardsException)
+
+      private[rewards] def getEpochNumber(
+        snapshotOrdinal: SnapshotOrdinal,
+        snapshotsPerEpoch: PosLong
+      ): Either[RewardsError, PosInt] =
+        snapshotOrdinal match {
+          case ordinal if ordinal == SnapshotOrdinal.MinValue => Right(1)
+          case ordinal =>
+            for (epoch <- 1 to config.epochs)
+              if (ordinal.value > (epoch - 1) * snapshotsPerEpoch && ordinal.value <= epoch * snapshotsPerEpoch)
+                return PosInt.from(epoch).left.map(RewardsException)
+            Left(LastEpochExceeded)
+
+        }
+
+      private[rewards] def createSharedDistribution(
+        snapshotHash: Hash,
+        snapshotReward: Amount,
+        addresses: NonEmptySet[Address]
+      ): F[Either[RewardsError, NonEmptyMap[Address, Amount]]] = {
+        val totalSpace = NonNegInt.unsafeFrom(addresses.length)
+        val (quotient, remainder) = snapshotReward.coerce.toLong /% totalSpace.toLong
+
+        val quotientDistribution = addresses.toNonEmptyList.map(_ -> quotient).toList.toMap.pure[F]
+
+        val remainderDistribution = if (remainder > 0) {
+          Random.scalaUtilRandomSeedInt(snapshotHash.value.toInt).flatMap { random =>
+            random.shuffleList(addresses.toNonEmptyList.toList).map { shuffled =>
+              shuffled.take(remainder.toInt).map(_ -> remainder / remainder).toMap
```

Comment:
`reminder / reminder` always has to be `1`, I think you can safely hardcode it 🙂

Reply:
Ah yes. It looked differently previously and I just changed the divisor to `remainder` and didn't notice that I ended up with `x/x` :D 

With `if (remainder > 0)` I can safely change it to `1` (actually without that check it would also throw on 0/0 anyway).